packages/expo-updates/android/src/main/java/expo/modules/updates/loader/Crypto.kt

```diff
@@ -104,4 +123,105 @@ object Crypto {
+    }
+
+    val keyId: String by lazy {
+      metadata?.get(CODE_SIGNING_METADATA_KEY_ID_KEY) ?: CODE_SIGNING_METADATA_DEFAULT_KEY_ID
+    }
+  }
+
+  fun createAcceptSignatureHeader(codeSigningConfiguration: CodeSigningConfiguration): String {
+    return Dictionary.valueOf(
+      mapOf(
+        CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_SIGNATURE to BooleanItem.valueOf(true),
+        CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_KEY_ID to StringItem.valueOf(codeSigningConfiguration.keyId),
+        CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_ALGORITHM to StringItem.valueOf(codeSigningConfiguration.algorithm.algorithmName)
+      )
+    ).serialize()
+  }
+
+  data class SignatureHeaderInfo(val signature: String, val keyId: String, val algorithm: CodeSigningAlgorithm)
+
+  fun parseSignatureHeader(signatureHeader: String?): SignatureHeaderInfo {
+    if (signatureHeader == null) {
+      throw Error("No expo-signature header specified")
+    }
+
+    val signatureMap = Parser(signatureHeader).parseDictionary().get()
+
+    val sigFieldValue = signatureMap[CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_SIGNATURE]
+    val keyIdFieldValue = signatureMap[CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_KEY_ID]
+    val algFieldValue = signatureMap[CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_ALGORITHM]
+
+    val signature = if (sigFieldValue is StringItem) {
+      sigFieldValue.get()
+    } else throw Error("Structured field $CODE_SIGNING_SIGNATURE_STRUCTURED_FIELD_KEY_SIGNATURE not found in expo-signature header")
+    val keyId = if (keyIdFieldValue is StringItem) {
+      keyIdFieldValue.get()
+    } else CODE_SIGNING_METADATA_DEFAULT_KEY_ID
+    val alg = if (algFieldValue is StringItem) {
+      algFieldValue.get()
+    } else null
+
+    return SignatureHeaderInfo(signature, keyId, CodeSigningAlgorithm.parseFromString(alg))
+  }
+
+  fun isSignatureValid(configuration: CodeSigningConfiguration, info: SignatureHeaderInfo, bytes: ByteArray): Boolean {
+    // check that the key used to sign the response is the same as the key embedded in the configuration
+    // TODO(wschurman): this may change for child certificates and development certificates
+    if (info.keyId != configuration.keyId) {
+      throw Error("Key with keyid=${info.keyId} from signature not found in client configuration")
+    }
+
+    // note that a mismatched algorithm doesn't fail early. it still tries to verify the signature with the
+    // algorithm specified in the configuration
+    if (info.algorithm != configuration.algorithm) {
+      Log.i(TAG, "Key with alg=${info.algorithm} from signature does not match client configuration algorithm, continuing")
+    }
+
+    return Signature.getInstance(
+      when (configuration.algorithm) {
+        CodeSigningAlgorithm.RSA_SHA256 -> "SHA256withRSA"
+      }
+    ).apply {
+      initVerify(configuration.embeddedCertificate.publicKey)
```

Comment:
Probably want to verify that configuration.keyId is root before using the embedded certificate

Reply:
The embedded certificate can have any keyId (specified in metadata). The only condition that needs to hold is that the keyId of the signature matches that of the embedded certificate (or one of the embedded certificates if we allow multiple in the future).

Reply:
I was looking for where that happens - is it correct that this code doesn't enforce that info.keyId (the signature's keyId) == embeddedCertificate.publicKey.keyId?

From a security standpoint this is fine since we always trust the embedded certificate but wondering if the spec allows for the signature's key ID to be on a different chain than the certificate that verifies the signature.

For instance, imagine a browser or other client with two trusted root certificates: ROOT_A and ROOT_B. The server sends a response that is signed by ROOT_A but declares `keyid=ROOT_B`.

Should the client accept or reject the signature?

Reply:
> I was looking for where that happens - is it correct that this code doesn't enforce that info.keyId (the signature's keyId) == embeddedCertificate.publicKey.keyId?

It does enforce this condition on L208, and it is also tested in `test_verifyCodeSigning_ThrowsWhenKeyDoesNotMatch`.

Reply:
I guess the next question is: how do we enforce that `configuration.keyId` matches `configuration.embeddedCertificate`? (Presumably this would happen at build time, but may want to do it at run time if it makes sense to do that.)

Reply:
Marking as resolved since the keyId is naturally separate from the certificate and not an intrinsic property of it.